py/modsys.c

```diff
@@ -54,7 +55,7 @@ const mp_print_t mp_sys_stdout_print = {&mp_sys_stdout_obj, mp_stream_write_adap
 #endif
 
 // version - Python language version that this implementation conforms to, as a string
-STATIC const MP_DEFINE_STR_OBJ(mp_sys_version_obj, "3.4.0");
+STATIC const MP_DEFINE_STR_OBJ(mp_sys_version_obj, "3.4.0; " MICROPY_BANNER_VERSION);
```

Comment:
Since this is just for informational purposes, could we drop the `3.4.0` here and just use `MICROPY_BANNER_VERSION` (and hopefully have one less copy of the string).

Reply:
Literal strings with a common suffix should de-duplicate. (Although we should verify that in compiled binary).

Reply:
You can see some builds have reduced code size with this PR, so I'm pretty sure gcc is de-duplicating the strings.

>  could we drop the 3.4.0 here

The reason to keep it is to match CPython, and also PyPy.  It's also a bit ambiguous as to whether it should be the CPython language version, or the implementation (uPy) version.

Here is what CPython docs say about `sys.implementation`:

----

sys.implementation

An object containing information about the implementation of the currently running Python interpreter. The following attributes are required to exist in all Python implementations.

name is the implementation’s identifier, e.g. 'cpython'. The actual string is defined by the Python implementation, but it is guaranteed to be lower case.

version is a named tuple, in the same format as [sys.version_info](https://docs.python.org/3/library/sys.html#sys.version_info). It represents the version of the Python implementation. This has a distinct meaning from the specific version of the Python language to which the currently running interpreter conforms, which sys.version_info represents. For example, for PyPy 1.8 sys.implementation.version might be sys.version_info(1, 8, 0, 'final', 0), whereas sys.version_info would be sys.version_info(2, 7, 2, 'final', 0). For CPython they are the same value, since it is the reference implementation.

----

From that it's clear that `sys.implementation.version` should be the uPy version.  And that `sys.version_info` should be the Python language version.

But it's not clear what `sys.version` should be.  The only hint is (as written at the top of this issue) that `sys.version_info` is the more programatic version of `sys.version`; the latter is a human readable string.

Reply:
I'd be happy to drop "3.4.0" from this string, it would make it simpler and neater, but I'm not sure that's the right thing to do.

Reply:
My point was that since is is just a human readable string, IMHO, humans should only care about the MicroPython version. The CPython compatible version (3.4.0) doesn't seem useful or relevant at this point in time since we implement many features from later versions of CPython.

Reply:
I agree with you.

But I want to do right by CPython.  Maybe we should ask them what's acceptable for this `sys.version` string.

Reply:
Even though the docs for `sys.version` say that one should not extract information from it, the `platform` module [does exactly that](https://github.com/python/cpython/blob/cb2b3c8d3566ae46b3b8d0718019e1c98484589e/Lib/platform.py#L964). So I agree leaving the python *implementation* version in `sys.version` seems like the right thing to do after all.

Reply:
Wow, that code for `_sys_version()` is insightful.  You can see that IronPython did put "IronPython" first in `sys.version`, but it looks like the Python language version was second, eg `IronPython 3.4.0`.

The information extracted by `_sys_version()` from the `sys.version` string is:
```
(name, version, branch, revision, buildno, builddate, compiler)
```

I found the following examples for this tuple:
```
('CPython', '2.7.13', '', '', 'default', 'Feb  6 2022 20:16:18', 'GCC 6.3.0 20170516')
('CPython', '2.7.18', '', '', 'default', 'Oct 10 2021 22:29:32', 'GCC 11.1.0')
('CPython', '3.5.3', '', '', 'default', 'Nov  4 2021 15:29:10', 'GCC 6.3.0 20170516')
('CPython', '3.10.4', '', '', 'main', 'Mar 23 2022 23:05:40', 'GCC 11.2.0')
('CPython', '3.10.0a7+', 'tags/v3.10.0a7-12-g489c36920e', '489c36920e', 'tags/v3.10.0a7-12-g489c36920e:489c36920e', 'Oct 21 2021 13:57:14', 'GCC 11.1.0')
('PyPy', '2.7.12', '', '', '5.6.0+dfsg-4', 'Nov 20 2016', '')
```

For MicroPython I guess we'd want something like this:
```
('MicroPython', '3.4.0', '', '', 'v1.18-377-geb9674822', '2022-04-23', 'GCC 11.2.0')
```
Which could correspond to a `sys.version` string like this:
```
"3.4.0; MicroPython v1.18-377-geb9674822 on 2022-04-23 [GCC 11.2.0]"
```
That's pretty much what is done in this PR, without the "GCC 11.2.0" bit.  We can leave the compiler string off, you can see that PyPy does not include it in the tuple above (although it does include it in its banner...).

In summary: keeping the "3.4.0" at the start seems reasonable, and matches the expectation of that `_sys_version()` parser (which gives us more information than the docs in this case).